packages/react/src/internal/usePresence.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright IBM Corp. 2025
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { useCallback, useEffect, useState, type RefObject } from 'react';
+
+export const usePresence = (
+  ref: RefObject<HTMLElement | null>,
+  isOpen: boolean
+) => {
+  const [exitState, setExitState] = useState<'idle' | 'active' | 'finished'>(
+    isOpen ? 'idle' : 'finished'
+  );
+  const isExiting = exitState === 'active';
+
+  if (!isOpen && exitState === 'idle') {
+    setExitState('active');
+  }
+  if (isOpen && exitState !== 'idle') {
+    setExitState('idle');
+  }
+
+  const handleAnimationEnd = useCallback(() => {
+    setExitState('finished');
+  }, []);
+
+  useEffect(() => {
+    if (!ref.current || !isExiting) return;
+    if (!('getAnimations' in ref.current)) {
+      handleAnimationEnd();
+      return;
+    }
+
+    const animations = ref.current.getAnimations();
```

Comment:
When applied to the Dialog, I think this will need to include the subtree option which will return animations on pseudo-elements like the `::backdrop`.

```suggestion
    const animations = ref.current.getAnimations({ subtree: true });
```

Reply:
Good to know. Thanks!